src/Microsoft.TemplateEngine.TemplateLocalizer.Core/TemplateStringExtractor.cs

```diff
@@ -0,0 +1,197 @@
+                                new StringFilteredTraversalRule("displayName"),
+                                // Include "description" of each choice.
+                                new StringFilteredTraversalRule("description"))))),
+                // Include "postActions" under the root, if they also comply with child rules.
+                new StringFilteredTraversalRule("postActions").WithChild(
+                    // Any post action in "postActions" array should be included. Skip none.
+                    new AllInclusiveTraversalRule().WithChildren(
+                        // Include "description" of the post action.
+                        new StringFilteredTraversalRule("description"),
+                        // Include "manualInstructions" of the post action, if they also comply with child rules.
+                        new RegexFilteredTraversalRule("manualInstructions").WithChild(
+                            // Include all the manual instructions in the array. Skip none.
+                            new AllInclusiveTraversalRule().WithChild(
+                                // Include "text" of the post action.
+                                new StringFilteredTraversalRule("text"))))));
+
+        public TemplateStringExtractor(JsonDocument document, ILoggerFactory? loggerFactory = null)
+        {
+            _jsonDocument = document;
+            _loggerFactory = loggerFactory ?? NullLoggerFactory.Instance;
+            _logger = _loggerFactory.CreateLogger<TemplateStringExtractor>();
+        }
+
+        public IReadOnlyList<TemplateString> ExtractStrings()
+        {
+            List<TemplateString> extractedStrings = new ();
+
+            TraversalArgs traversalArgs = new(
+                    identifierPrefix: string.Empty,
+                    keyPrefix: string.Empty,
+                    rules: new List<TraversalRule>() { _documentRootTraversalRule },
+                    extractedStrings,
+                    extractedStringIds: new HashSet<string>());
+
+            TraverseJsonElements(
+                _jsonDocument.RootElement,
+                elementName: string.Empty,
+                localizationKey: string.Empty,
+                traversalArgs);
+            return extractedStrings;
+        }
+
+        private void TraverseJsonElements(
+            JsonElement element,
+            string elementName,
+            string localizationKey,
+            TraversalArgs args)
+        {
+            using IDisposable? _ = _logger.BeginScope(elementName);
+            List<TraversalRule> complyingRules = args.Rules.Where(r => r.AllowsTraversalOfIdentifier(elementName)).ToList();
+
+            if (complyingRules.Count == 0)
+            {
+                // This identifier was filtered out.
+                _logger.LogInformation(
+                    "The following element in the template.json will not be included in the localizations" +
```

Comment:
minor: consider caching the string used in the `_logger.LogInformation` call and just "+" `elementName`.  I'm unsure if this string is created on every such call.

Reply:
The string literals in the first argument will be combined at compile time and interned at runtime. The second argument will be needed for #3058, since we don't know how the string will be formatted in different languages..